addons/edi_stock/models/edi_move_cancellation_report_document.py

```diff
@@ -0,0 +1,171 @@
+    """
+
+    _name = "edi.document.stock.cancellation"
+    _inherit = "edi.move.report.document"
+    _description = "Stock Cancellations"
+
+    _edi_move_report_via = "edi_stock_cancellation_doc_id"
+
+    @api.model
+    def move_report_domain(self, _doc):
+        """
+        Override to only include cancelled moves for picking types which have
+        u_enable_pick_cancellation_email_notifs enabled.
+        _edi_move_report_via functionality is kept to ensure moves are not reported on twice
+        """
+        StockPickingType = self.env["stock.picking.type"]
+        picking_types_to_consider = StockPickingType.search(
+            [("u_enable_pick_cancellation_email_notifs", "=", True)]
+        )
+        domain = [
+            ("state", "=", "cancel"),
+            (self._edi_move_report_via, "=", False),
+            ("picking_id.picking_type_id", "in", picking_types_to_consider.ids),
+        ]
+        return domain
+
+    @api.model
+    def move_report_list(self, doc, moves):
+        """
+        Inherit to group moves by sale & picking_type to result
+        in a cancellation record for each unique combination
+        """
+        res = super().move_report_list(doc, moves)
+        grouped_moves = []
+        for sales_order, moves_by_sale in res.groupby(lambda move: move.sale_line_id.order_id):
+            # Moves for the same sale could be for multiple pick types, which means
+            # they could have different templates configured
+            for _picking_type, moves_by_template in moves_by_sale.groupby(
+                lambda move: move.picking_type_id
+            ):
+                # If we have no sale associated, then we need to further group by pick partner
+                # to ensure emails will be sent separately per partner
+                if not sales_order:
+                    for _pick_parner, moves_by_pick_partner in moves_by_sale.groupby(
+                        lambda move: move.picking_id.partner_id
+                    ):
+                        grouped_moves.append(moves_by_pick_partner)
+                else:
+                    grouped_moves.append(moves_by_template)
+        return grouped_moves
+
+    @api.model
+    def execute(self, doc):
+        """For each line, send an email using the template set on the line"""
+        super().execute(doc)
+        for record in doc.move_cancellation_ids:
+            _logger.info(
+                f"Sending email for Move Cancellation document: {doc.name}, Stock Cancellation: {record.name}"
+            )
+            record.email_template_id.send_mail(record.id, force_send=True)
+
+
+class EdiStockCancellationRecord(models.Model):
+    """Stock Cancellation record"""
+
+    _name = "edi.record.stock.cancellation"
+    _inherit = "edi.move.report.record"
+    _description = "Stock Cancellation"
+
+    order_reference = fields.Char(
+        string="Order Reference", required=True, readonly=True, index=True
```

Comment:
General we overuse indexes, we don't need them on _every_ field. It's something that needs looking at another time though and not here... just wanted to moan...